src/ui/PackageDependencyProvider.ts

```diff
@@ -148,38 +155,130 @@ export class PackageDependenciesProvider implements vscode.TreeDataProvider<Tree
+                    const version = this.dependencyDisplayVersion(dependency);
+                    const packagePath = this.dependencyPackagePath(
+                        dependency,
+                        folderContext.folder.fsPath
+                    );
+                    const location = dependency.packageRef.location;
+                    return new PackageNode(
+                        dependency.packageRef.identity,
+                        packagePath,
+                        location,
+                        version,
+                        type
+                    );
+                }) ?? []
+        );
+    }
+
+    /**
+     * * Returns a set of all dependencies that are in use in the workspace.
+     * 1. local/remote/edited dependency has remote/edited dependencies, Package.resolved covers them
+     * 2. remote/edited dependency has a local dependency, the local dependency must have been declared in root Package.swift
+     * 3. local dependency has a local dependency, traverse it and find the local dependencies only recursively
+     * 4. pins include all remote and edited packages for 1, 2
+     */
+    private async getInUseDependencies(
+        workspaceState: WorkspaceState,
+        folderContext: FolderContext
+    ): Promise<Set<string>> {
+        const localDependencies = await this.getAllLocalDependencySet(
+            workspaceState,
+            folderContext
+        );
+        const remoteDependencies = this.getRemoteDependencySet(folderContext);
+
+        // merge the two sets of dependencies
+        const inUseDependencies = new Set<string>();
+        localDependencies.forEach(dependency => inUseDependencies.add(dependency));
+        remoteDependencies.forEach(dependency => inUseDependencies.add(dependency));
+        return inUseDependencies;
+    }
+
+    private getRemoteDependencySet(folderContext: FolderContext | undefined): Set<string> {
+        return new Set<string>(folderContext?.swiftPackage.resolved?.pins.map(pin => pin.identity));
+    }
+
+    private async getAllLocalDependencySet(
+        workspaceState: WorkspaceState,
+        folderContext: FolderContext
+    ): Promise<Set<string>> {
+        const rootDependencies = folderContext?.swiftPackage.dependencies ?? [];
+        const showingDependencies = new Set<string>();
+        await this.getChildLocalDependencySet(
+            rootDependencies,
+            workspaceState?.object.dependencies ?? [],
+            folderContext?.folder.fsPath,
+            showingDependencies
+        );
+
+        return showingDependencies;
+    }
+
+    /**
+     * tranverse only local dependency tree
+     * @param dependencies depdencies of current node
+     * @param workspaceStateDependencies all dependencies in workspace-state.json
+     * @param showingDependencies result of dependencies that are showing in the tree
+     * @returns
+     */
+    private async getChildLocalDependencySet(
```

Comment:
Is it worth having `this.getChildLocalDependencySet` return a Set instead of passing a Set in to be filled. When you are parsing children of children you could combine the sets. It makes for slightly cleaner code.

Reply:
Maybe converting to a stack version?